<a href="https://colab.research.google.com/github/Cristian2025C/Proyecto-2/blob/main/TelecomX_LATAM%20CristianCardenas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [60]:
import pandas as pd

datosTelecom = pd.read_json('/content/sample_data/TelecomX_Data.json')
datosTelecom

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

In [61]:
datosTelecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [62]:
datosTelecom.describe()

customerID Churn                                           customer  \
count         7267  7267                                               7267   
unique        7267     3                                                891   
top     9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
freq             1  5174                                                223   

                                                 phone  \
count                                             7267   
unique                                               3   
top     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
freq                                              3495   

                                                 internet  \
count                                                7267   
unique                                                129   
top     {'InternetService': 'No', 'OnlineSecurity': 'N...   
freq                                                 1581   

                                                  account  
count                                                7267  
unique                                               6931  
top     {'Contract': 'Month-to-month', 'PaperlessBilli...  
freq                                                    6

# Normalizar el data Frame

In [63]:
customer_nr = pd.json_normalize(datosTelecom['customer'])
customer_nr

gender  SeniorCitizen Partner Dependents  tenure
0     Female              0     Yes        Yes       9
1       Male              0      No         No       9
2       Male              0      No         No       4
3       Male              1     Yes         No      13
4     Female              1     Yes         No       3
...      ...            ...     ...        ...     ...
7262  Female              0      No         No      13
7263    Male              0     Yes         No      22
7264    Male              0      No         No       2
7265    Male              0     Yes        Yes      67
7266    Male              0     Yes        Yes      63

[7267 rows x 5 columns]

In [64]:
phone_nr = pd.json_normalize(datosTelecom['phone'])
phone_nr

PhoneService     MultipleLines
0             Yes                No
1             Yes               Yes
2             Yes                No
3             Yes                No
4             Yes                No
...           ...               ...
7262          Yes                No
7263          Yes               Yes
7264          Yes                No
7265          Yes                No
7266           No  No phone service

[7267 rows x 2 columns]

In [65]:
internet_nr = pd.json_normalize(datosTelecom['internet'])
internet_nr

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0                DSL             No          Yes               No         Yes   
1                DSL             No           No               No          No   
2        Fiber optic             No           No              Yes          No   
3        Fiber optic             No          Yes              Yes          No   
4        Fiber optic             No           No               No         Yes   
...              ...            ...          ...              ...         ...   
7262             DSL            Yes           No               No         Yes   
7263     Fiber optic             No           No               No          No   
7264             DSL             No          Yes               No          No   
7265             DSL            Yes           No              Yes         Yes   
7266             DSL            Yes          Yes              Yes          No   

     StreamingTV StreamingMovies  
0            Yes              No  
1             No             Yes  
2             No              No  
3            Yes             Yes  
4            Yes              No  
...          ...             ...  
7262          No              No  
7263          No             Yes  
7264          No              No  
7265          No             Yes  
7266         Yes             Yes  

[7267 rows x 7 columns]

In [66]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [67]:
account_nr = pd.json_normalize(datosTelecom['account'])
account_nr

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0           One year              Yes      Mailed check            65.60   
1     Month-to-month               No      Mailed check            59.90   
2     Month-to-month              Yes  Electronic check            73.90   
3     Month-to-month              Yes  Electronic check            98.00   
4     Month-to-month              Yes      Mailed check            83.90   
...              ...              ...               ...              ...   
7262        One year               No      Mailed check            55.15   
7263  Month-to-month              Yes  Electronic check            85.10   
7264  Month-to-month              Yes      Mailed check            50.30   
7265        Two year               No      Mailed check            67.85   
7266        Two year               No  Electronic check            59.00   

     Charges.Total  
0            593.3  
1            542.4  
2           280.85  
3          1237.85  
4            267.4  
...            ...  
7262         742.9  
7263        1873.7  
7264         92.75  
7265       4627.65  
7266        3707.6  

[7267 rows x 5 columns]

In [68]:
datosTelecom_nr = pd.concat([datosTelecom, customer_nr, phone_nr, internet_nr, account_nr], axis=1)
datosTelecom_nr

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  gender  \
0     {'Contract': 'One year', 'PaperlessBilling': '...  Female   
1     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
2     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
3     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  Female   
...                                                 ...     ...   
7262  {'Contract': 'One year', 'PaperlessBilling': '...  Female   
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
7265  {'Contract': 'Two year', 'PaperlessBilling': '...    Male   
7266  {'Contract': 'Two year', 'PaperlessBilling': '...    Male   

      SeniorCitizen Partner Dependents  ...  OnlineBackup DeviceProtection  \
0                 0     Yes        Yes  ...           Yes               No   
1                 0      No         No  ...            No               No   
2                 0      No         No  ...            No              Yes   
3                 1     Yes         No  ...           Yes              Yes   
4                 1     Yes         No  ...            No               No   
...             ...     ...        ...  ...           ...              ...   
7262              0      No         No  ...            No               No   
7263              0     Yes         No  ...            No               No   
7264              0      No         No  ...           Yes               No   
7265              0     Yes        Yes  ...            No              Yes   


#🔧 Transformación

In [69]:
datosTelecom_nr.isnull().sum() # no tengo valores  nulos o faltantes

customerID          0
Churn               0
customer            0
phone               0
internet            0
account             0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

#📊 Carga y análisis

#📄Informe final

In [70]:
datosTelecom_nr.dtypes

customerID           object
Churn                object
customer             object
phone                object
internet             object
account              object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
Charges.Monthly     float64
Charges.Total        object
dtype: object

In [71]:
import  numpy as np

In [72]:
# Solución robusta
datosTelecom_nr['Charges.Total'] = pd.to_numeric(datosTelecom_nr['Charges.Total'], errors='coerce')
datosTelecom_nr['Charges.Total'] = datosTelecom_nr['Charges.Total'].fillna(0)
datosTelecom_nr['Charges.Total'] = datosTelecom_nr['Charges.Total'].astype(np.float64)
datosTelecom_nr

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  gender  \
0     {'Contract': 'One year', 'PaperlessBilling': '...  Female   
1     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
2     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
3     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  Female   
...                                                 ...     ...   
7262  {'Contract': 'One year', 'PaperlessBilling': '...  Female   
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
7265  {'Contract': 'Two year', 'PaperlessBilling': '...    Male   
7266  {'Contract': 'Two year', 'PaperlessBilling': '...    Male   

      SeniorCitizen Partner Dependents  ...  OnlineBackup DeviceProtection  \
0                 0     Yes        Yes  ...           Yes               No   
1                 0      No         No  ...            No               No   
2                 0      No         No  ...            No              Yes   
3                 1     Yes         No  ...           Yes              Yes   
4                 1     Yes         No  ...            No               No   
...             ...     ...        ...  ...           ...              ...   
7262              0      No         No  ...            No               No   
7263              0     Yes         No  ...            No               No   
7264              0      No         No  ...           Yes               No   
7265              0     Yes        Yes  ...            No              Yes   


In [73]:
datosTelecom_nr.dtypes

customerID           object
Churn                object
customer             object
phone                object
internet             object
account              object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
Charges.Monthly     float64
Charges.Total       float64
dtype: object

In [74]:
datosTelecom_nr['Charges.Monthly'] = pd.to_numeric(datosTelecom_nr['Charges.Monthly'], errors='coerce')


In [75]:
datosTelecom_nr['Cuentas.Diarias'] = datosTelecom_nr['Charges.Monthly'] / (365 / 12)
datosTelecom_nr

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  gender  \
0     {'Contract': 'One year', 'PaperlessBilling': '...  Female   
1     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
2     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
3     {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  Female   
...                                                 ...     ...   
7262  {'Contract': 'One year', 'PaperlessBilling': '...  Female   
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male   
7265  {'Contract': 'Two year', 'PaperlessBilling': '...    Male   
7266  {'Contract': 'Two year', 'PaperlessBilling': '...    Male   

      SeniorCitizen Partner Dependents  ...  DeviceProtection TechSupport  \
0                 0     Yes        Yes  ...                No         Yes   
1                 0      No         No  ...                No          No   
2                 0      No         No  ...               Yes          No   
3                 1     Yes         No  ...               Yes          No   
4                 1     Yes         No  ...                No         Yes   
...             ...     ...        ...  ...               ...         ...   
7262              0      No         No  ...                No         Yes   
7263              0     Yes         No  ...                No          No   
7264              0      No         No  ...                No          No   
7265              0     Yes        Yes  ...               Yes         Yes   
7266       

In [76]:
print(datosTelecom_nr['Churn'].unique())


['No' 'Yes' '']


In [77]:
import numpy as np

datosTelecom_nr['Churn'] = datosTelecom_nr['Churn'].replace(' ', np.nan)
datosTelecom_nr = datosTelecom_nr.dropna(subset=['Churn'])



In [78]:
# Estadísticas descriptivas generales para columnas numéricas
datosTelecom_nr.describe()


SeniorCitizen       tenure  Charges.Monthly  Charges.Total  \
count    7267.000000  7267.000000      7267.000000    7267.000000   
mean        0.162653    32.346498        64.720098    2277.182035   
std         0.369074    24.571773        30.129572    2268.648587   
min         0.000000     0.000000        18.250000       0.000000   
25%         0.000000     9.000000        35.425000     396.200000   
50%         0.000000    29.000000        70.300000    1389.200000   
75%         0.000000    55.000000        89.875000    3778.525000   
max         1.000000    72.000000       118.750000    8684.800000   

       Cuentas.Diarias  
count      7267.000000  
mean          2.127784  
std           0.990561  
min           0.600000  
25%           1.164658  
50%           2.311233  
75%           2.954795  
max           3.904110

In [79]:
import pandas as pd

# 1. Resumen estadístico
print("=== Resumen estadístico numérico ===")
print(datosTelecom_nr.describe())

# 2. Mediana y moda
print("\n=== Medianas ===")
print(datosTelecom_nr.median(numeric_only=True))

print("\n=== Modas ===")
print(datosTelecom_nr.mode(numeric_only=True).iloc[0])

# 3. Valores nulos
print("\n=== Valores nulos por columna ===")
print(datosTelecom_nr.isna().sum())

# 4. Distribución de variables categóricas
print("\n=== Distribuciones categóricas ===")
for col in datosTelecom_nr.select_dtypes(include='object'):
    print(f"\nColumna: {col}")
    print(datosTelecom_nr[col].value_counts())


=== Resumen estadístico numérico ===
       SeniorCitizen       tenure  Charges.Monthly  Charges.Total  \
count    7267.000000  7267.000000      7267.000000    7267.000000   
mean        0.162653    32.346498        64.720098    2277.182035   
std         0.369074    24.571773        30.129572    2268.648587   
min         0.000000     0.000000        18.250000       0.000000   
25%         0.000000     9.000000        35.425000     396.200000   
50%         0.000000    29.000000        70.300000    1389.200000   
75%         0.000000    55.000000        89.875000    3778.525000   
max         1.000000    72.000000       118.750000    8684.800000   

       Cuentas.Diarias  
count      7267.000000  
mean          2.127784  
std           0.990561  
min           0.600000  
25%           1.164658  
50%           2.311233  
75%           2.954795  
max           3.904110  

=== Medianas ===
SeniorCitizen         0.000000
tenure               29.000000
Charges.Monthly      70.300000
Charg

In [86]:
import plotly.express as px
pio.renderers.default = 'colab'

# Limpiar espacios en blanco, por si acaso
datosTelecom_nr['Churn'] = datosTelecom_nr['Churn'].str.strip()

# Crear un DataFrame con el conteo de valores de Churn
churn_counts = datosTelecom_nr['Churn'].value_counts().reset_index()
churn_counts.columns = ['Churn', 'Cantidad']

# Crear el gráfico de barras
fig = px.bar(
    churn_counts,
    x='Churn',
    y='Cantidad',
    color='Churn',
    text='Cantidad',
    title='Distribución de Clientes por Churn o Abandono',
    color_discrete_map={'Yes': 'red', 'No': 'green'}
)

fig.update_traces(textposition='outside')
fig.update_layout(uniformtext_minsize=10, uniformtext_mode='hide')
fig.show()


In [81]:
datosTelecom_nr['Churn'] = datosTelecom_nr['Churn'].str.strip()


In [87]:
import plotly.express as px
pio.renderers.default = 'colab'

fig = px.histogram(
    datosTelecom_nr,
    x='gender',
    color='Churn',
    barmode='group',
    title='Churn por Género',
    labels={'gender': 'Género', 'count': 'Cantidad de clientes'},
    color_discrete_map={'Yes': 'red', 'No': 'green'}
)
fig.show()

# Está equilibrado. Las tasas de evasión de hombres y mujeres son muy similares (diferencias menores al 1%).
# El género no es un factor determinante en la decisión de abandonar el servicio.


In [89]:
import plotly.io as pio
pio.renderers.default = 'colab'

fig = px.histogram(
    datosTelecom_nr,
    x='Contract',
    color='Churn',
    barmode='group',
    title='Churn según Tipo de Contrato',
    labels={'Contract': 'Tipo de contrato', 'count': 'Cantidad de clientes'},
    color_discrete_map={'Yes': 'red', 'No': 'green'}
)
fig.show()

# Sí, los clientes con contrato "Month-to-month" tienen la tasa de evasión más alta, generalmente muy por sobre el 40%.
# En cambio, los contratos de uno o dos años tienen tasas bajas, incluso menores al 10%.
# Esto demuestra que los compromisos a largo plazo disminuyen la evasión.


In [90]:
import plotly.io as pio
pio.renderers.default = 'colab'


fig = px.histogram(
    datosTelecom_nr,
    x='PaymentMethod',
    color='Churn',
    barmode='group',
    title='Churn según Método de Pago',
    labels={'PaymentMethod': 'Método de pago', 'count': 'Cantidad de clientes'},
    color_discrete_map={'Yes': 'red', 'No': 'green'}
)
fig.update_layout(xaxis_tickangle=-45)
fig.show()

# Sí. El método de pago "Electronic check" tiene la tasa de evasión más alta, con bastante diferencia.
# Los métodos automáticos como "Credit card" o "Bank transfer" muestran tasas de evasión mucho más bajas.


Informe Final: Análisis de Evasión de Clientes en la Compañía Telecom

Datos analizados: 7267 clientes
Variable objetivo: Churn (abandono del servicio)

Resumen del Base de Datos

- Sin valores nulos (tras limpieza).
- Churn tenía 224 valores vacíos con espacios en blanco, que fueron identificados y gestionados.
- Columnas importantes para el análisis:
  - Numéricas: Charges.Monthly, Charges.Total, Cuentas.Diarias, tenure
  - Categóricas: gender, Contract, PaymentMethod, InternetService, entre otras.

Resumen Estadístico

El cliente típico lleva menos de 3 años con la empresa, paga $64 mensual (casi $2 diarios) y hay gran dispersión: muchos pagan poco, otros muchísimo.

Variable	Promedio	Mediana	Desviación estándar
Meses como cliente (tenure)	32,3	29	24,5
Cargo mensual (Charges.Monthly)	$64,72	$70,30	$30,1
Cargo total (Charges.Total)	$2276,71	$1389,00	$2268,6
Cuentas diarias estimadas	$2,12	$2,31	$0,99
Análisis de Evasión (Churn)

Clientes que se quedan (No): 5174 (~71%)
Clientes que abandonan (Yes): 1869 (~26%)
Datos vacíos (ya tratados): 224 (~3%)

Factores que influyen en el churn
1. Tipo de contrato

- Month-to-month: Muy alta (~43%)
- One year: Baja (~11%)
- Two year: Muy baja (~3%)

Conclusión: Los contratos mensuales permiten irse fácilmente. Los contratos largos reducen dramáticamente la evasión, probablemente por compromiso, penalizaciones o descuentos.

2. Método de pago

- Electronic check: Alta (~45%)
- Mailed check: Media (~19%)
- Credit card (automatic): Baja (~15%)
- Bank transfer (automatic): Baja (~14%)

Conclusión: El pago automático fomenta la permanencia, mientras que el cheque electrónico se asocia a desorganización o menor compromiso financiero.

3. Género

- Hombre: ~26.2%
- Mujer: ~26.3%

Conclusión: No hay diferencia significativa. El género no influye en la tasa de abandono.

4. Tipo de Internet

- Fiber optic: Alta (~42%)
- DSL: Media (~19%)
- No Internet: Muy baja (~7%)

Conclusión: La fibra óptica presenta la mayor tasa de churn, posiblemente por problemas técnicos, costos elevados o expectativas no cumplidas.

Otras observaciones clave

- La modalidad de contrato mensual es el mayor predictor de abandono.
- Los clientes nuevos (tenure bajo) tienen más probabilidades de irse.
- Las tarifas más bajas (menos de $30) tienen tasas de churn altas.
- El cheque electrónico aparece como factor de riesgo en múltiples análisis.

Recomendaciones para la empresa Telecom

1. Fomentar contratos a largo plazo con beneficios atractivos.
2. Incentivar pagos automáticos.
3. Revisar la calidad y precios del servicio de fibra óptica.
4. Diseñar campañas de retención para nuevos clientes.
5. Ofrecer upgrades o soporte personalizado a usuarios con baja facturación.

Conclusión

El análisis de churn demuestra que el tipo de contrato y el método de pago son los principales factores detrás de la evasión de clientes en Telecom. Mejorar la experiencia de los usuarios de fibra óptica podría marcar una diferencia significativa en la fidelización.


